### PR TITLE
Add `system76/launch_1` keyboard features

### PR DESCRIPTION
**Pull request by jecassis**
## Description

Implements the following changes for System76's Launch keyboard (launch_1):

- Disable mechanical locking support (`LOCKING_SUPPORT_ENABLE`) and switch state consistency with keyboard LED state (`LOCKING_RESYNC_ENABLE`) by default.
- Use `TOTAL_EEPROM_BYTE_COUNT` from AVR library for total size for portability.
- Start custom EEPROM region using `EECONFIG_SIZE` (mirrors how VIA does the same).
- Add README for default keymap.
- Remove custom Bootmagic Lite function and enable QMK's implementation in the Makefile.
- Update comments on almost all keyboard and keymap files.
- Allow removing System76 EC (i.e., configurator) functionality through a Makefile option (still enabled by default).
- Add header file for System76 EC to ease sharing between various System76 keyboards.
- Add key combination to unlock the keyboard and flash via System76 Firmware Updater (independent of QMK bootloader features).
- Update USB "multiplexer" and document code behavior in substantial detail. Check changes for consistency with the USB Type-C Port Controller Interface Specification.
- Enable removal of custom RGB matrix effects for space-saving (as implemented in #17478).
- Correct a couple of minor comment typos in core QMK files.
- Integrate and format changes from #17889 and #17037.
- Update C formatting (using `clang-format`).
- Update copyright year.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR

- None

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/bootmagic/bootmagic_lite.c
Comment (drashna) [inline quantum/bootmagic/bootmagic_lite.c:20]: These changes should be pulled out and added to a seperate PR

Comment (jecassis) [reply quantum/bootmagic/bootmagic_lite.c]: Removed these changes from this PR. I leave it to the core maintainers to correct at your convenience.

File: keyboards/system76/launch_1/keymaps/default/readme.md
Comment (drashna) [inline keyboards/system76/launch_1/keymaps/default/readme.md:3]: Please replace this with markdown.

**Review by drashna — APPROVED** [review]

File: keyboards/system76/launch_1/rules.mk
Comment (fauxpark) [inline keyboards/system76/launch_1/rules.mk:5]:
```suggestion
# MCU name
MCU = atmega32u4

# CPU frequency divided by two since AVR is at 3.3 V
F_CPU = 8000000

# External oscillator is 16 MHz
F_USB = 16000000

# Bootloader selection
BOOTLOADER = atmel-dfu
```

File: keyboards/system76/launch_1/rules.mk
Comment (fauxpark) [inline keyboards/system76/launch_1/rules.mk:7]:
```suggestion
# Build options
#   change yes to no to disable
#
```

File: keyboards/system76/launch_1/rules.mk
Comment (fauxpark) [inline keyboards/system76/launch_1/rules.mk:15]:
```suggestion
RAW_ENABLE = yes            # Raw HID commands (used by Keyboard Configurator)
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/system76/launch_1/keymaps/default/readme.md
Comment (jecassis) [reply keyboards/system76/launch_1/keymaps/default/readme.md]: GitHub Markdown cannot add custom widths, but I don't care to argue. Expressed the HTML indirectly in Markdown sans the width constraint. It'll look disproportionately blown up on wide screens.

File: keyboards/system76/launch_1/rules.mk
Comment (jecassis) [reply keyboards/system76/launch_1/rules.mk]: I get you want it to be exactly the same on all of the "rules" Makefiles. If you want to enforce this, may I recommend you do so in an automated fashion.

File: keyboards/system76/launch_1/rules.mk
Comment (jecassis) [reply keyboards/system76/launch_1/rules.mk:23]: Also added the extra knobs as requested, but don't understand why as these 3 are explicitly set to "no" in the common features Makefile. The `show_full_features` target returns:
```text
Hardware Options:
  AUDIO_ENABLE                   = no               # file
  BACKLIGHT_ENABLE               = no               # file
  RGBLIGHT_ENABLE                = no               # file
```

File: keyboards/system76/launch_1/rules.mk
Comment (jecassis) [reply keyboards/system76/launch_1/rules.mk]: Done, again, hesitantly, because there is a good reason for doing it the other way. Basically, it enables tracking down the places where inbuilt assumptions of no trailing whitespaces are located for these special variables. So that if a user inadvertently adds one it will not break a build like it does today. Do not assume we all use GNU Make.

Comment (fauxpark) [reply keyboards/system76/launch_1/rules.mk]: I think it's actually a fairly reasonable assumption. For instance, [the CLI prefers `gmake` before `make`](https://github.com/qmk/qmk_firmware/blob/master/lib/python/qmk/commands.py#L22) when compiling boards, and on macOS, the Homebrew formula [depends on GNU `make`](https://formulae.brew.sh/formula/make). The FreeBSD install script also specifies `gmake`.

**Comment by PeterFalken** [comment]
@jecassis, FYI PR #17478 was merged recently.

**Comment by PeterFalken** [comment]
@jecassis , I think that you only need to rebase this PR with master and that should take care of the conflicts.

**Review by PeterFalken — APPROVED** [review]

File: keyboards/system76/system76_ec.c
Comment (drashna) [inline keyboards/system76/system76_ec.c:31]: Most of these includes shouldn't be needed, if you have quantum.h included.

File: keyboards/system76/launch_1/config.h
Comment (drashna) [inline keyboards/system76/launch_1/config.h:21]: this shouldn't ben necessary

Comment (jecassis) [reply keyboards/system76/launch_1/config.h]: Unfortunately, it is necessary. Unfamiliar with the build dependencies well enough to implement a better workaround. Tried with AVR LIBC 5.4.0 and 7.3.0. Build error output:
```log
Assembling: platforms/avr/xprintf.S                                                                 [ERRORS]
 | 
 | /usr/lib/avr/include/stdint.h: Assembler messages:
 | /usr/lib/avr/include/stdint.h:125: Error: unknown opcode `typedef'
...
 | /usr/lib/avr/include/stdint.h:282: Error: unknown opcode `typedef'
 | quantum/eeconfig.h:92: Error: unknown opcode `_bool'
 | quantum/eeconfig.h:95: Error: unknown opcode `void'
 | quantum/eeconfig.h:104: Error: unknown opcode `uint8_t'
...
 | quantum/eeconfig.h:134: Error: unknown opcode `void'
 | 
gmake[1]: *** [builddefs/common_rules.mk:359: .build/obj_system76_launch_1_default/platforms/avr/xprintf.o] Error 1
```

Comment (drashna) [reply keyboards/system76/launch_1/config.h:21]: 5.4 is old enough, that it doesn't support a lot.  It shouldn't be used, period. 

We specifically recommend 8.x for a reason.  And a big part of that is supported feature set of the compilers.  

So, yeah, this shouldn't be necessary, at all.

File: keyboards/system76/launch_1/config.h
Comment (drashna) [inline keyboards/system76/launch_1/config.h:52]: why is the formatting being undone? 

Running `qmk format-c` on these files will undo the changes you've added here.

File: keyboards/system76/launch_1/keymaps/default/readme.md
Comment (drashna) [inline keyboards/system76/launch_1/keymaps/default/readme.md:3]:
```suggestion
![launch_1 Base Layout Image](https://i.imgur.com/7RTC4iVh.png "launch_1 Base Layout")
```

**Review by sigprof — CHANGES_REQUESTED** [review]
Apparently I forgot to click “Submit review” at the end, so my comment was left pending here for several months :man_facepalming:

File: keyboards/system76/launch_1/config.h
Comment (jecassis) [reply keyboards/system76/launch_1/config.h]: Reran ClangFormat on `config.h`.

File: keyboards/system76/launch_1/config.h
Comment (jecassis) [reply keyboards/system76/launch_1/config.h]: Same behavior with 7.3.0 as with 8.3.0, 9.2.0, 10.1.0, 11.1.0 and 12.1.0. It does not appear to be a compiler feature issue. As mentioned, there is something abstracted into the build system that is manifesting in these side effects. Consequently, hardcoded the EEPROM size variable to work around the need to include any files in `config.h`.

**Comment by jecassis** [comment]
This has been ready to merge for a while. Just waiting for final maintainer approval. Can you please take a look and advise?

**Pull request closed without merging**
